packages/eslint-plugin/src/rules/prefer-destructuring.ts

```diff
@@ -0,0 +1,232 @@
+      const typeChecker = program.getTypeChecker();
+      const baseRules = baseRule.create(context);
+      const rules =
+        leftNode.type === AST_NODE_TYPES.Identifier &&
+        leftNode.typeAnnotation === undefined
+          ? baseRules
+          : baseRule.create(noFixContext(context));
+      if (
+        'typeAnnotation' in leftNode &&
+        leftNode.typeAnnotation !== undefined &&
+        !enforceForDeclarationWithTypeAnnotation
+      ) {
+        return;
+      }
+
+      if (
+        rightNode != null &&
+        isArrayLiteralIntegerIndexAccess(rightNode) &&
+        rightNode.object.type !== AST_NODE_TYPES.Super
+      ) {
+        const tsObj = esTreeNodeToTSNodeMap.get(rightNode.object);
+        const objType = typeChecker.getTypeAtLocation(tsObj);
+        if (!isTypeIterableType(objType, typeChecker)) {
+          if (
+            !enforceForRenamedProperties ||
+            !getNormalizedEnabledType(reportNode.type, 'object')
+          ) {
+            return;
+          }
+          context.report({
+            node: reportNode,
+            messageId: 'preferDestructuring',
+            data: { type: 'object' },
+          });
+          return;
+        }
+      }
+
+      if (reportNode.type === AST_NODE_TYPES.AssignmentExpression) {
+        rules.AssignmentExpression(reportNode);
+      } else {
+        rules.VariableDeclarator(reportNode);
+      }
+    }
+
+    function getNormalizedEnabledType(
+      nodeType:
+        | AST_NODE_TYPES.VariableDeclarator
+        | AST_NODE_TYPES.AssignmentExpression,
+      destructuringType: 'array' | 'object',
+    ): boolean | undefined {
+      if ('object' in enabledTypes || 'array' in enabledTypes) {
+        return enabledTypes[destructuringType];
+      }
+      return enabledTypes[nodeType as keyof typeof enabledTypes][
+        destructuringType as keyof (typeof enabledTypes)[keyof typeof enabledTypes]
+      ];
+    }
+  },
+});
+
+type Context = TSESLint.RuleContext<MessageIds, Options>;
+
+function noFixContext(context: Context): Context {
+  const customContext: {
+    report: Context['report'];
+  } = {
+    report: (descriptor): void => {
+      context.report({
+        ...descriptor,
+        fix: undefined,
+      });
+    },
+  };
+
+  // we can't directly proxy `context` because its `report` property is non-configurable
+  // and non-writable. So we proxy `customContext` and redirect all
+  // property access to the original context except for `report`
+  return new Proxy<Context>(customContext as typeof context, {
+    get(target, path, receiver): unknown {
+      if (path !== 'report') {
+        return Reflect.get(context, path, receiver);
+      }
+      return Reflect.get(target, path, receiver);
+    },
+  });
+}
+
+function isTypeIterableType(
+  type: ts.Type,
+  typeChecker: ts.TypeChecker,
+): boolean {
+  if (isTypeAnyType(type)) {
+    return true;
+  }
+  if (!type.isUnion()) {
```

Comment:
[Cleanup] Interestingly, `require-await.ts` already has some code around this pattern of checking all types for a well known symbol property:

https://github.com/typescript-eslint/typescript-eslint/blob/3cdf5c93163036fdc60325129d9746ff57574336/packages/eslint-plugin/src/rules/require-await.ts#L120-L122

https://github.com/typescript-eslint/typescript-eslint/blob/3cdf5c93163036fdc60325129d9746ff57574336/packages/eslint-plugin/src/rules/require-await.ts#L237-L242

What do you think about making a shared helper to deduplicate some of the code? Think it's worth it?

(not a blocker at all)

Reply:
I think we can't use `expandUnionOrIntersectionType` is not suitable for `prefer-destructuring`.
If we use it, it returns same results for types both like `object & unknown[]` and `object | unknown[]` though former is iterable and latter is not iterable.
I need to inspect `require-await` to make a shared helper. If `expandUnionOrIntersectionType` works properly for `require-await`, maybe we don't need shared helper. Or not, I think making a shared helper looks reasonable and the helper will fix a bug of `require-await`.
I suppose that shared helper should be added in another PR because it will change the behavior `require-await` a bit.